From 6170e80372a8974c9a421c59940685bdb9825f4c Mon Sep 17 00:00:00 2001
From: Lorenzo Villani <lvillani@develer.com>
Date: Thu, 15 Oct 2015 16:34:17 +0200
Subject: [PATCH] Add support for gzipped tar outer packages

---
 libopkg/opkg_archive.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/libopkg/opkg_archive.c b/libopkg/opkg_archive.c
index 12312e6..aa998f2 100644
--- a/libopkg/opkg_archive.c
+++ b/libopkg/opkg_archive.c
@@ -502,6 +502,24 @@ static struct archive *open_outer(const char *filename)
         goto err_cleanup;
     }
 
+    r = archive_read_support_filter_gzip(outer);
+    if (r == ARCHIVE_WARN) {
+        /* libarchive returns ARCHIVE_WARN if the filter is provided by
+         * an external program.
+         */
+        opkg_msg(INFO, "Gzip support provided by external program.\n");
+    } else if (r != ARCHIVE_OK) {
+        opkg_msg(ERROR, "Gzip format not supported.\n");
+        goto err_cleanup;
+    }
+
+    r = archive_read_support_format_tar(outer);
+    if (r != ARCHIVE_OK) {
+        opkg_msg(ERROR, "Tar format not supported: %s\n",
+                 archive_error_string(outer));
+        goto err_cleanup;
+    }
+
     r = archive_read_open_filename(outer, filename, EXTRACT_BUFFER_LEN);
     if (r != ARCHIVE_OK) {
         opkg_msg(ERROR, "Failed to open package '%s': %s\n", filename,
-- 
2.6.1

