From d691f42a49deaaa132c51c942a791af5127b87be Mon Sep 17 00:00:00 2001
From: Giovanni Bajo <rasky@develer.com>
Date: Thu, 17 Sep 2015 17:09:59 +0200
Subject: [PATCH] Change default name for second-stage bootloader
Organization: Develer S.r.l.

---
 Config.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Config.in b/Config.in
index 96d8630..29da9ec 100644
--- a/Config.in
+++ b/Config.in
@@ -263,7 +263,7 @@ endmenu
 config CONFIG_IMAGE_NAME
 	string
 	default "Image" if CONFIG_LINUX_IMAGE
-	default "u-boot.bin" if CONFIG_LOAD_UBOOT
+	default "barebox.bin" if CONFIG_LOAD_UBOOT
 	default "softpack.bin" if CONFIG_LOAD_64KB || CONFIG_LOAD_4MB || CONFIG_LOAD_1MB
 	  
 config CONFIG_DEBUG
-- 
1.9.1

